### PR TITLE
Removed RenderHarvester, RenderLandingCraft and RenderUnit

### PR DESCRIPTION
**Pull request by reaperrr**
With this PR, `RenderUnit` and its childs are finally removed.

Replaces `RenderHarvester` with `WithHarvesterAnimation`, and `RenderLandingCraft` with `WithLandingCraftAnimation`, and removes all remaining usages of `RenderUnit` as well as the trait itself.

Additionally fixes the namespace of `WithAttackAnimation` and the name of `WithHarvestOverlay` (used by d2k and TS harvesters).

### CODE REVIEW

**Comment by chrisforbes** [comment]
Untested, but looks OK. :+1:

**Comment by reaperrr** [comment]
Updated (RA landing craft now uses ```WithSpriteBody``` instead of ```WithFacingSpriteBody```)

**Comment by reaperrr** [comment]
As talked about on irc, split ```WithHarvesterAnimation``` into ```WithHarvestAnimation``` and ```WithDockingAnimation```.

File: OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs [hunk 1]
```diff
@@ -9,64 +9,66 @@
...
 
 namespace OpenRA.Mods.Common.Traits
 {
-	[Desc("Displays an overlay whenever resources are harvested by the actor.")]
-	class WithHarvestAnimationInfo : ITraitInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>
+	public class WithHarvestAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>, Requires<HarvesterInfo>
 	{
-		[Desc("Sequence name to use")]
-		[SequenceReference] public readonly string Sequence = "harvest";
+		public readonly string[] PrefixByFullness = { "" };
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs]: Needs a `[SequenceReference(null, true)]`

File: OpenRA.Mods.D2k/Traits/Sandworm.cs [hunk 2]
```diff
@@ -32,6 +32,12 @@ class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>
 		[Desc("Name of the sequence that is used when the actor is idle or moving (not attacking).")]
 		public readonly string IdleSequence = "idle";
 
+		[Desc("Name of the sequence that is used when the actor is attacking.")]
+		public readonly string MouthSequence = "mouth";
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Sandworm.cs]: Missing `[SequenceReference]`.

File: OpenRA.Mods.D2k/Traits/Sandworm.cs [hunk 3]
```diff
@@ -32,6 +32,12 @@ class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>
...
+
+		[Desc("Name of the sequence that is used when the actor is burrowed.")]
+		public readonly string BurrowedSequence = "burrowed";
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Sandworm.cs]: Missing `[SequenceReference]`.

File: OpenRA.Mods.D2k/Traits/Sandworm.cs [hunk 4]
```diff
@@ -32,6 +32,12 @@ class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>
 		[Desc("Name of the sequence that is used when the actor is idle or moving (not attacking).")]
 		public readonly string IdleSequence = "idle";
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Sandworm.cs]: Missing `[SequenceReference]`.

File: mods/ra/sequences/vehicles.yaml [hunk 5]
```diff
@@ -14,48 +14,38 @@ truk:
 	icon: trukicon
 
 harv:
-	idle:
+	idle: harvempty
```
Comment (Mailaender) [inline mods/ra/sequences/vehicles.yaml]: I think this needs to be `empty-harv` now.

Comment (Mailaender) [inline mods/ra/sequences/vehicles.yaml]: Or not.

**Comment by reaperrr** [comment]
One regression I noticed myself is that harvest animations in RA and TD now play too fast. Might be an issue with the PlayCustomAnimation[Repeating] implementations in ```WithSpriteBody```, or something else.

Increasing the Tick value of the sequence actually makes things worse. I suspect the animation is restarted for every Harvested() op now rather than simply being looped while harvesting, and since LoadTicksPerBale is 4 Ticks by default and the animation is 8 frames long in RA for example, they don't sync anymore.

 I haven't figured out how to fix that yet, any help would be appreciated.

**Comment by reaperrr** [comment]
Added the missing SequenceReferences and rebased.

File: OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs [hunk 6]
```diff
@@ -9,64 +9,67 @@
...
+				if (!IsModifying && wsb.DefaultAnimation.CurrentSequence.Name != NormalizeHarvesterSequence(self, wsb.Info.Sequence))
+					wsb.DefaultAnimation.ReplaceAnim(NormalizeHarvesterSequence(self, wsb.Info.Sequence));
+			}
+		}
 
-			visible = true;
-			anim.PlayThen(info.Sequence, () => visible = false);
+		public void Harvested(Actor self, ResourceType resource)
+		{
+			if (!string.IsNullOrEmpty(Info.HarvestSequence) && wsb.DefaultAnimation.HasSequence(NormalizeHarvesterSequence(self, Info.HarvestSequence)))
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs]: I'd assume this check only needs to pass if IsModifying is false.

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs]: That fixes the issue described above, thanks.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs]: Glad I could've helped.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 7]
```diff
@@ -1206,6 +1206,136 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
...
 				}
 
+				// Removed RenderUnit
+				if (engineVersion < 20150630)
+				{
+					// Renamed WithHarvestAnimation to WithHarvestOverlay
+					if (node.Key == "WithHarvestAnimation")
+						node.Key = "WithHarvestOverlay";
+
+					// Replaced RenderLandingCraft with With*Body + WithLandingCraftAnimation.
```
Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Nit: the star - "With*Body"

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 8]
```diff
@@ -1206,6 +1206,136 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
...
+							rlc.Value.Nodes.RemoveAll(n => rsNodes.Contains(n));
+							rlc.Value.Nodes.RemoveAll(n => wfsbNodes.Contains(n));
+						}
+
+						var rrlc = node.Value.Nodes.FirstOrDefault(n => n.Key.StartsWith("-RenderLandingCraft"));
+						if (rrlc != null)
+							rrlc.Key = "-WithLandingCraftAnimation";
+					}
+
+					// Replaced RenderHarvester with WithFacingSpriteBody + WithHarvestAnimation.
```
Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Nit: + WithDockingAnimation

**Comment by penev92** [comment]
Something tells me this will introduce several 
```
+	RenderSprites:
+	WithFacingSpriteBody:
+	AutoSelectionSize:
```
that you just removed in your YAML cleanup PRs.

File: OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs [hunk 9]
```diff
@@ -15,24 +15,34 @@ namespace OpenRA.Mods.Common.Activities
...
+			wsb = self.Trait<WithSpriteBody>();
+			wda = self.Trait<WithDockingAnimation>();
 		}
 
 		public override Activity OnStateDock(Actor self)
 		{
-			ru.PlayCustomAnimation(self, "dock", () => ru.PlayCustomAnimationRepeating(self, "dock-loop"));
+			var wha = self.TraitOrDefault<WithHarvestAnimation>();
+			if (wha != null)
+				wha.IsModifying = true;
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs]: Why would the dock sequence care about `WithHarvestAnimation` ?

File: OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs [hunk 10]
```diff
@@ -15,24 +15,34 @@ namespace OpenRA.Mods.Common.Activities
...
 			return this;
 		}
 
 		public override Activity OnStateUndock(Actor self)
 		{
-			ru.PlayCustomAnimationBackwards(self, "dock", () => dockingState = State.Complete);
+			wsb.PlayCustomAnimationBackwards(self, wda.Info.DockSequence, () => dockingState = State.Complete);
+			var wha = self.TraitOrDefault<WithHarvestAnimation>();
+			if (wha != null && dockingState == State.Complete)
+				wha.IsModifying = false;
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs]: Same question. Also if it's a must, you can cache the trait. But in any case, this looks too suspicious.

File: OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs (hunk 9 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs]: If ```WithHarvestAnimation.IsModifying``` isn't set to true, the docking anim will be overridden by the WHA fullness modifier.

Comment (penev92) [inline OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs]: Ugh that is ... not nice.
Can you add a comment to explain this workaround please.

**Comment by reaperrr** [comment]
It would be better if #8514 goes in first so I can rebase against it and make sure I won't introduce new duplication.

**Comment by penev92** [comment]
#8514 was merged.

**Comment by reaperrr** [comment]
Rebased.

**Comment by Mailaender** [comment]
I like how you spotted more hard-coded sequence and exposed those to testing. :+1: 

One crash remaining:

```
Loading mod: d2k
Exception of type `System.InvalidOperationException`: Actor does not have trait of type `OpenRA.Mods.Common.Traits.WithDockingAnimation`
  at OpenRA.TraitDictionary+TraitContainer`1[OpenRA.Mods.Common.Traits.WithDockingAnimation].Get (UInt32 actor) [0x0003e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/TraitDictionary.cs:150 
  at OpenRA.TraitDictionary.Get[WithDockingAnimation] (OpenRA.Actor actor) [0x00013] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/TraitDictionary.cs:96 
  at OpenRA.Actor.Trait[WithDockingAnimation] () [0x0000d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:212 
  at OpenRA.Mods.Common.Activities.SpriteHarvesterDockSequence..ctor (OpenRA.Actor self, OpenRA.Actor refinery, Int32 dockAngle, Boolean isDragRequired, WVec dragOffset, Int32 dragLength) [0x0001e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Activities/SpriteHarvesterDockSequence.cs:25 
  at OpenRA.Mods.Common.Traits.Refinery.DockSequence (OpenRA.Actor harv, OpenRA.Actor self) [0x0001b] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:76 
  at OpenRA.Mods.Common.Traits.Refinery.OnDock (OpenRA.Actor harv, OpenRA.Mods.Common.Activities.DeliverResources dockOrder) [0x00051] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:134 
  at OpenRA.Mods.Common.Activities.DeliverResources.Tick (OpenRA.Actor self) [0x001cc] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Activities/DeliverResources.cs:69 
  at OpenRA.Traits.Util.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00017] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Traits/Util.cs:83 
  at OpenRA.Actor.Tick () [0x00010] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:138 
  at OpenRA.World.Tick () [0x000de] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/World.cs:311 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bf] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:486 
  at OpenRA.Game.LogicTick () [0x0004e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:510 
  at OpenRA.Game.Loop () [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:634 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:674 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:39 
AL lib: (EE) alc_cleanup: 1 device not closed

```

**Comment by reaperrr** [comment]
Updated.

**Comment by penev92** [comment]
Oh wait! I knew there was a reason I wanted that interface to notify on docking/undocking!
We can have `WithDockAnimation` handle
```
wsb.PlayCustomAnimation(self, wda.Info.DockSequence, () => wsb.PlayCustomAnimationRepeating(self, wda.Info.DockLoopSequence));
```
and the corresponding undock, and clean up `SpriteHarvesterDockSequence` even further.

**Comment by reaperrr** [comment]
Actually, wouldn't it make even more sense to revert most of the last batch of changes and use ```INotifyDocking```?

**Comment by reaperrr** [comment]
> But I guess I've asked for enough out of this PR already, so not going to hold it up on that.

As it won't require yaml changes, I'd prefer to leave that for a follow-up PR, those last-minute changes always carry some risk of overlooking something.

**Comment by Mailaender** [comment]
We didnt't use `INotfifyDocking` on purpose https://github.com/OpenRA/OpenRA/pull/5797#issuecomment-48013075.

**Comment by Mailaender** [comment]
I just noted another problem. Now the Red Alert harvest sequence is gone after the first unload.

**Comment by pchote** [comment]
My objections to the interface we purely within the context of that earlier pr.  I have no opinion over whether it is suitable here (because i havent read the code yet).

File: OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs [hunk 11]
```diff
@@ -9,64 +9,79 @@
...
+		public void Docked()
+		{
+			IsModifying = true;
+		}
 
-			visible = true;
-			anim.PlayThen(info.Sequence, () => visible = false);
+		public void Undocked()
+		{
+			IsModifying = true;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs]: This must be the problem. It should be set to false, but only AFTER the undock animation has finished, so simply setting ```IsModifying = false``` here won't do the trick as it breaks/prematurely aborts the undock animation.

**Comment by reaperrr** [comment]
Updated.

**Comment by penev92** [comment]
There is a visual glitch with the landing craft's closing animation.
Other than that :+1:

**Comment by reaperrr** [comment]
You mean that it starts to close when moving and then restarts the animation when entering the next cell? I noticed that too, but that's on bleed as well. I'll try to fix it, but I might need help (again).

**Comment by reaperrr** [comment]
Rebased.

**Comment by Mailaender** [comment]
> Warning: Sprite image delphi from actor delphi of faction allies does not define sequence shoot from field AttackSequence of OpenRA.Mods.Common.Traits.WithInfantryBodyInfo

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b1cf0b3dd7047bd3493c371242dee2123290dfee...b2833cda66a0b94d8617cc228e1e480138c5a16a)

**Pull request merged**